crates/paimon/src/file_index/file_index_format.rs

```diff
@@ -0,0 +1,514 @@
+            buffer.put_i32_le(index_info.len() as i32);
+            // ColumnInfo，offset = headLength
+            for (index_name, IndexInfo { start_pos, length }) in index_info {
+                buffer.put_u16_le(index_name.len() as u16);
+                buffer.put_slice(index_name.as_bytes());
+                let adjusted_start = if *start_pos == EMPTY_INDEX_FLAG {
+                    EMPTY_INDEX_FLAG
+                } else {
+                    *start_pos + head_length as i32
+                };
+                buffer.put_i32_le(adjusted_start);
+                buffer.put_i32_le(*length);
+            }
+        }
+
+        // Redundant length for future compatibility
+        buffer.put_i32_le(0);
+        Ok(buffer.freeze())
+    }
+
+    fn calculate_head_length(
+        &self,
+        body_info: &HashMap<String, HashMap<String, IndexInfo>>,
+    ) -> crate::Result<usize> {
+        // Magic + Version + HeadLength + ColumnNumber + RedundantLength
+        let base_length = 8 + 4 + 4 + 4 + 4;
+        let mut total_length = base_length;
+
+        for (column_name, index_info) in body_info {
+            // Column name length + actual column name length
+            total_length += 2 + column_name.len();
+            // IndexTypeSize (index number)
+            total_length += 4;
+
+            for index_name in index_info.keys() {
+                // Index name length + actual index name length
+                total_length += 2 + index_name.len();
+                // start_pos (4 bytes) + length (4 bytes)
+                total_length += 8;
+            }
+        }
+
+        Ok(total_length)
+    }
+
+    pub fn from_file_io(file_io: FileIO, path: &str) -> crate::Result<Self> {
+        let output_file = file_io.new_output(path)?;
+        Ok(Self::new(output_file))
+    }
+}
+
+#[derive(Debug)]
+struct IndexInfo {
+    start_pos: i32,
+    length: i32,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum Version {
+    V1,
+}
+
+impl From<i32> for Version {
+    fn from(version: i32) -> Self {
+        match version {
+            1 => Version::V1,
+            _ => panic!("Unsupported file index version: {}", version),
+        }
+    }
+}
+
+impl From<Version> for i32 {
+    fn from(version: Version) -> Self {
+        match version {
+            Version::V1 => 1,
+        }
+    }
+}
+
+pub struct FileIndexFormatReader {
+    reader: Box<dyn FileRead>,
+    header: HashMap<String, HashMap<String, IndexInfo>>,
+}
+
+impl FileIndexFormatReader {
+    pub async fn new(input_file: InputFile) -> crate::Result<Self> {
+        let reader = input_file.reader().await?;
+        let mut file_reader = Self {
+            reader: Box::new(reader),
+            header: HashMap::new(),
+        };
+        file_reader.read_header().await?;
+        Ok(file_reader)
+    }
+
+    async fn read_header(&mut self) -> crate::Result<()> {
+        // Magic (8 bytes)
+        let mut buffer = self.reader.read(0..8).await?;
```

Comment:
Hi, reading 8B of data from storage services is very slow and costly. It's better to optimize by reading the entire header and then decoding it. We can try reading the first 1MiB to see if it contains the entire header and read more if it doesn't.

Reply:
Indeed, that is better. I will figure it out.